### PR TITLE
Update upwind cfl

### PR DESCRIPTION
**Pull request by alessiofumagalli**
- update upwind and coupling
- use the new param interface
- add test for upwind-coupling

### CODE REVIEW

**Comment by keileg** [comment]
@alessiofumagalli , @IvarStefansson Travis fails; could you please have a look?

File: src/porepy/grids/grid_bucket.py [hunk 1]
```diff
@@ -728,6 +728,85 @@ def eliminate_node(self, node):
 
 #------------------------------------------------------------------------------#
 
+    def loop_nodes(self, fct):
```
Comment (keileg) [inline src/porepy/grids/grid_bucket.py]: 1) Can't this be achieved by values = [fct(g, d) for g, d in gb]
2) If you still want to keep the function, the name should reflect that you evaluate a function on all nodes

File: src/porepy/grids/grid_bucket.py [hunk 2]
```diff
@@ -728,6 +728,85 @@ def eliminate_node(self, node):
...
+            i[idx], j[idx] = self.nodes_prop([g_l, g_h], 'node_number')
+            values[idx] = fct(g_h, g_l, data_h, data_l, data)
+            idx += 1
+
+        # Upper triangular matrix
+        return sps.coo_matrix((values, (i, j)), (self.size(), self.size()))
+
+#------------------------------------------------------------------------------#
+
+    def loop(self, fct_nodes, fct_edges):
```
Comment (keileg) [inline src/porepy/grids/grid_bucket.py]: Function name needs to change.

File: src/porepy/grids/grid_bucket.py [hunk 3]
```diff
@@ -728,6 +728,85 @@ def eliminate_node(self, node):
...
+        """
+        i = np.zeros(self.graph.number_of_edges(), dtype=int)
+        j = np.zeros(i.size, dtype=int)
+        values = np.zeros(i.size)
+
+        # Loop over the edges of the graph (pair of connected nodes)
+        idx = 0
+        for e, data in self.edges_props():
+            g_l, g_h = self.sorted_nodes_of_edge(e)
+            data_l, data_h = self.node_props(g_l), self.node_props(g_h)
```
Comment (keileg) [inline src/porepy/grids/grid_bucket.py]: I really don't like such assignments.

File: examples/example8/aa_test_advection_diffusion.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import numpy as np
```
Comment (alessiofumagalli) [inline examples/example8/aa_test_advection_diffusion.py]: correct

File: src/porepy/grids/grid_bucket.py (hunk 3 above)
Comment (alessiofumagalli) [inline src/porepy/grids/grid_bucket.py]: what do you mean? It was already there..

File: src/porepy/grids/grid_bucket.py (hunk 1 above)
Comment (alessiofumagalli) [inline src/porepy/grids/grid_bucket.py]: 1. we need to be consistent with the node ordering
2. ok

File: src/porepy/grids/grid_bucket.py (hunk 2 above)
Comment (alessiofumagalli) [inline src/porepy/grids/grid_bucket.py]: ok

File: examples/example8/aa_test_advection_diffusion.py (hunk 4 above)
Comment (keileg) [inline examples/example8/aa_test_advection_diffusion.py]: Fair enough.

File: src/porepy/grids/grid_bucket.py (hunk 3 above)
Comment (keileg) [inline src/porepy/grids/grid_bucket.py]: Well, then I won't blame you. Nevertheless, it is hard to read code when you have to count the number of commas on each side of the equality sign. I much prefer an extra line or two.

**Comment by keileg** [comment]
You have my sympathy.

Please do not merge this branch directly - it needs a thorough squeeze of commits when Travis becomes satisfied.

**Pull request merged**